Mantis #46252 Some elements in roleassignment table are clickable eventhough permissions are not sufficient

ingoj opened the pull request:
This PR offers a possible fix for mantis 46252

ingoj committed:
avoid permission errors

this fix checks if the current user has proper permissions for roles and context. Links are only rendered if permissions for the container for a local role are granted
The path display has the same problem about permissions, but a fix for this is needed in Services/Tree (will be suggested soon)

ingoj committed:
Codestyle fixes

ingoj committed:
More Style fixes

kergomard commented:
Thank you very much for the PR @ingoj !

This generally looks right, There is one error (I think):

The check [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10496/files#diff-c5773eb8746881c4f4327d09a98a52ff32c378e129e46562bfcf827395966879R148-R149) should not be for either or, but it should be `visible,read` to check for both.

...and there are a few things that  are important to me:
- [ ] Please do not use the `$DIC` in the middle of a class. You should create a class-property `access` in the constructor of the class and assign `ilAccess` to it in the constructor. You can then use it with `$this->access` in the function.
- [ ] I do really not like places where we need an `else` and even less if we have nested `if`s. For me this is a sign that the function is too long/complex. Please move the corresponding parts to their own functions with early returns, so that you never need more than one level of conditions and never need an `else`. The two function should return the value that the variables "ROLE" and "CONTEXT" are set to respectively and they should be called `buildRoleVariableValue(string $title)` and `buildContextVariableValue(int $ref_id)`.

Thanks again and best,
@kergomard

ingoj committed:
Streamline Code

add class property access
restructure if checks for role and context links
introduce variable $role to handle link generation analogously to $context

ingoj committed:
cs-fix

ingoj committed:
cs fix

ingoj commented:
Thanks for yor Review, @kergomard !
I Updated my repo accordingly. If you have further changes, feel free to proceed.
I did not update the if check for read/visible, since the links work porperly if one of the permissions is granted.

ingoj committed:
completely remove $DIC from fillRow

ingoj commented:
@kergomard ,

today I saw that I had omitted some of the planned Changes, I updated this a couple of minutes ago.

kergomard commented:
Thank you very much for the PR @ingoj  .

I had to add a little change that was not worth the trouble of a round trip so I added this change manually with https://github.com/ILIAS-eLearning/ILIAS/commit/11920d280c4785e4abd0839a277c6e11817b9fd2 and picked it to ILIAS 10, ILIAS 11, and trunk.

Best,
@kergomard
